emoji/src/main/java/com/vanniktech/emoji/EmojiHandler.java

```diff
@@ -1,34 +1,75 @@
 package com.vanniktech.emoji;
 
 import android.content.Context;
+import android.support.v4.util.Pair;
 import android.text.Spannable;
 import com.vanniktech.emoji.emoji.Emoji;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.vanniktech.emoji.EmojiHandler.SpanRangeList.SPAN_NOT_FOUND;
 
 final class EmojiHandler {
-  static void addEmojis(final Context context, final Spannable text, final int emojiSize) {
-    final EmojiSpan[] spans = text.getSpans(0, text.length(), EmojiSpan.class);
+  private EmojiHandler() {
+    throw new AssertionError("No instances.");
+  }
 
-    for (final EmojiSpan oldSpan : spans) {
-      text.removeSpan(oldSpan);
-    }
+  static void addEmojis(final Context context, final Spannable text, final int emojiSize) {
+    final SpanRangeList existingSpanRanges = new SpanRangeList(text);
+    final EmojiManager emojiManager = EmojiManager.getInstance();
+    int index = 0;
 
-    int i = 0;
+    while (index < text.length()) {
+      final int existingSpanEnd = existingSpanRanges.spanEnd(index);
 
-    final EmojiManager instance = EmojiManager.getInstance();
+      if (existingSpanEnd == SPAN_NOT_FOUND) {
+        final int nextSpanStart = existingSpanRanges.nextSpanStart(index);
+        final int searchRange = nextSpanStart == SPAN_NOT_FOUND ? text.length() : nextSpanStart;
+        final Emoji found = emojiManager.findEmoji(text.subSequence(index, searchRange));
 
-    while (i < text.length()) {
-      final Emoji found = instance.findEmoji(text.subSequence(i, text.length()));
+        if (found != null) {
+          text.setSpan(new EmojiSpan(context, found.getResource(), emojiSize), index, index + found.getLength(),
+                  Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
 
-      if (found != null) {
-        text.setSpan(new EmojiSpan(context, found.getResource(), emojiSize), i, i + found.getLength(), Spannable.SPAN_INCLUSIVE_EXCLUSIVE);
-        i += found.getLength();
+          index += found.getLength();
+        } else {
+          index++;
+        }
       } else {
-        i++;
+        index += existingSpanEnd - index;
       }
     }
   }
 
-  private EmojiHandler() {
-    throw new AssertionError("No instances.");
+  static final class SpanRangeList {
+    static final int SPAN_NOT_FOUND = -1;
+
+    private final List<Pair<Integer, Integer>> spanRanges = new ArrayList<>();
```

Comment:
Might be better to write our own Pair class that uses primitive `int` instead of `Integer` otherwise you'll have plenty of boxing and unboxing which can also cause bad performance